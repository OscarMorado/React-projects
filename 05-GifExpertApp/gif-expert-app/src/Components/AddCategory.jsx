import { useState } from "react"
import { miembros } from "../data/Members";

export const AddCategory = ({onNewCategory}) => {

    const[ inputValue, setInputValue] = useState('Honoka');
    const[ errorMessage, setErrorMessage] = useState("");

    const getMiembro = (nombre) => { //Valida que el nombre ingresado en el input coincida con uno de los nombres del archivo 'miembros'

        const nameFormat = (str) =>
            str
                .split(" ")
                .map((nombre) => nombre.charAt(0).toUpperCase() + nombre.slice(1).toLowerCase())
                .join(" ")

        // return Object.values(miembros[0]).some((miembro) => miembro.some((miembro) => miembro.includes(nameFormat(nombre))));  //Devuelve true si se ingresa una sola parte del nombre (nombre o apellido). Devuelve un bool, por lo que se comentó para permitir que se devuelva el miembro como tal

        for (const miembrosGrupo of Object.values(miembros[0])) {
            for (const miembro of miembrosGrupo) {
                if (miembro.includes(nameFormat(nombre))) {
                    return miembro; //Devuelve el nombre completo
                }
            }
        }

        return null;
    };

    const onInputChange = (event) => {
        setInputValue(event.target.value);
    }

    const onSubmit = (event) => {
        event.preventDefault();
        
        const nombreCompleto = getMiembro(inputValue);
        if(!getMiembro(nombreCompleto)){
            setErrorMessage("Ingresa el nombre de una idol de Love Live!");
            console.log(errorMessage);
            return;
        }
        
        onNewCategory(nombreCompleto);
        setInputValue('');
        setErrorMessage("");
    }

    return (
        <form onSubmit = {onSubmit}>
            <p>{errorMessage}</p>
            <input
                type = "text"
                placeholder = "Busca otro miembro!"
                value = {inputValue}
                onChange ={onInputChange}
            />
        </form>
    )
}


/* 
    Si bien este componente solamente debería agregar el nuevo valor a la categoría, se le añadió más código para validar la entrada de dicha categoría antes de añadir el valor
    Esto se podría mover a un nuevo componente que se encargue de validar la entrada, sin embargo, se dejará así por cuestiones de tiempo 

*/