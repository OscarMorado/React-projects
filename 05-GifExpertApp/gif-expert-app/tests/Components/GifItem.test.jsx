import {render, screen} from '@testing-library/react';
import { GifItem } from '../../src/Components/GifItem';


describe('Pruebas en GifItem', () => {

    const title = 'Honoka Kousaka';
    const url = 'https://lovelive.com';

    test('Debe hacer match con el snapshot ', () => {
        const {container} = render(<GifItem title = {title} url = {url}/>);
        expect(container).toMatchSnapshot();
    });
  
    test('Debe mostrar la imagen con el URL y ALT indicado ', () => {
        render(<GifItem title = {title} url = {url}/>);
        screen.debug();
        expect(screen.getByRole('img').alt);
    })

    test('Debe mostrar el titulo en el componente ', () => {  //Solo funcionará con el title del <p>, ya que el otro title está en un <alt>
        render(<GifItem title = {title} url = {url}/>);
        expect(screen.getByText(title)).toBeTruthy();
    })
    
})
