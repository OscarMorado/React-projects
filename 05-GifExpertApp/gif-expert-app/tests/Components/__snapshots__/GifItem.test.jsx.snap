// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifItem Debe hacer match con el snapshot  1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Honoka Kousaka"
      src="https://lovelive.com"
    />
    <p>
      Honoka Kousaka
    </p>
  </div>
</div>
`;
